src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -225,14 +229,42 @@ class GcsRpcClient {
   }
 
   void Reconnect() {
+    absl::MutexLock lock(&mutex_);
     if (get_server_address_) {
-      auto address = get_server_address_();
-      Init(address.first, address.second, client_call_manager_);
+      std::pair<std::string, int> address;
+      int index = 0;
+      for (; index < RayConfig::instance().ping_gcs_rpc_server_max_retries(); ++index) {
+        address = get_server_address_();
+        if (Ping(address.first, address.second, 100)) {
+          RAY_LOG(INFO) << "Target rpc server " << address.first << ":" << address.second
+                        << " is valid.";
+          break;
+        }
+        usleep(RayConfig::instance().ping_gcs_rpc_server_interval_milliseconds() * 1000);
+      }
+
+      if (index < RayConfig::instance().ping_gcs_rpc_server_max_retries()) {
+        Init(address.first, address.second, client_call_manager_);
+        if (reconnected_callback_) {
+          reconnected_callback_();
+        }
+      } else {
+        RAY_LOG(FATAL) << "Target rpc server " << address.first << ":" << address.second
+                       << " is invalid.";
+      }
     }
   }
 
+  absl::Mutex mutex_;
+
   ClientCallManager &client_call_manager_;
   std::function<std::pair<std::string, int>()> get_server_address_;
+  /// This function is triggered when RPC request is called.
+  /// We will ping the address of the GCS server, and if it is valid, this function will
+  /// be called. Currently, we will subscribe to GCS again in this function. If the
+  /// function call succeeds but the connection of GCS server fails, it will try again and
+  /// subscription function needs to be idempotent.
```

Comment:
/// The callback that will be called when we reconnect to GCS server.
/// Currently, we use this function to reestablish subscription to GCS. 
/// Note, we use `ping` to detect whether the reconnection is successful. If the ping succeeds but the
/// RPC connection fails, this function might be called called again. So it needs to be idempotent.

Reply:
fixed